### PR TITLE
Features/patch

### PR DESCRIPTION
**Pull request by brendankowitz**
## Description
Pulling across Patch changes from https://github.com/microsoft/fhir-server/pull/1875 and adding a few improvements and tests.

## Related issues
Addresses #1268

## Testing
Adds E2E tests
Adds manual .http test file

- [x] Test for changing a choice type
- [x] JSON "Test" operation test
- [x] Test for properties with an extension
- [x] Test for array items with an extension 
- [ ] Spec test cases https://www.hl7.org/fhir/test-cases.zip

## FHIR Team Checklist
- [x] **Update the title** of the PR to be succinct and less than 50 characters
- [x] **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- [x] Tag the PR with the type of update: **Bug**, **Dependencies**, **Enhancement**, or **New-Feature**
- [ ] Tag the PR with **Azure API for FHIR** if this will release to the managed service
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Feature

### CODE REVIEW

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/PatchResourceHandler.cs [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+                resource = node.ToTypedElement(_modelInfoProvider.StructureDefinitionSummaryProvider);
+                (string path, object result)[] postState = _immutableProperties.Select(x => (path: x, result: resource.Scalar(x))).ToArray();
+
+                if (!preState.Zip(postState).All(x => x.First.path == x.Second.path && string.Equals(x.First.result?.ToString(), x.Second.result?.ToString(), StringComparison.Ordinal)))
+                {
+                    throw new RequestNotValidException(Core.Resources.PatchImmutablePropertiesIsNotValid);
+                }
+
+                // Persist
+                Resource resourcePoco = resource.ToPoco<Resource>();
```
Comment (Ivanidzo4ka) [inline src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/PatchResourceHandler.cs:115]: This need to be validated before persistence.

**Comment by Ivanidzo4ka** [comment]
>     internal class CapabilityStatementBuilder : ICapabilityStatementBuilder

Need to add 
```
"patchFormat":[
 "application/json-patch+json"
],
```

---
Refers to: src/Microsoft.Health.Fhir.Core/Features/Conformance/CapabilityStatementBuilder.cs:27 in 92af89c. [](commit_id = 92af89c97939287e16903657338b0b693c9dcdf7, deletion_comment = False)

**Review by Ivanidzo4ka — APPROVED** [review]
:shipit:

**Comment by CaitlinV39** [comment]
@brendankowitz - Do you think were okay to merge this?

**Pull request merged**